### PR TITLE
flamenco: size out worst case space for spad-backed borrowed accounts

### PR DESCRIPTION
**Pull request by yufeng-jump**
- Include space for spad allocator metadata block, i.e. fd_spad_t
- Include alignment padding on each individual borrowed account

### CODE REVIEW

File: src/flamenco/runtime/tests/fd_exec_sol_compat.c [hunk 1]
```diff
@@ -64,7 +63,7 @@ sol_compat_wksp_init( void ) {
   wksp = fd_wksp_new_anonymous( FD_SHMEM_NORMAL_PAGE_SZ, 65536UL * 8UL, fd_shmem_cpu_idx( fd_shmem_numa_idx( cpu_idx ) ), "wksp", 0UL );
   assert( wksp );
 
-  spad_mem = fd_wksp_alloc_laddr( wksp, FD_SPAD_ALIGN, spad_mem_max, 3 ); /* 1.28 GB */
+  spad_mem = fd_wksp_alloc_laddr( wksp, FD_SPAD_ALIGN, fd_spad_footprint( 128UL * fd_ulong_align_up( FD_ACC_TOT_SZ_MAX, FD_ACCOUNT_META_ALIGN ) ), 3 ); /* 1342191744 B */
```
Comment (topointon-jump) [inline src/flamenco/runtime/tests/fd_exec_sol_compat.c:66]: When this is merged, could you notify @nlgripto that our fuzzing memory requirements have slightly increased? 🙏

File: src/flamenco/runtime/fd_runtime.c [hunk 2]
```diff
@@ -1823,7 +1823,7 @@ fd_runtime_finalize_txns_tpool( fd_exec_slot_ctx_t *         slot_ctx,
         fd_acc_mgr_view( txn_ctx->acc_mgr, txn_ctx->funk_txn, &txn_ctx->accounts[0], borrowed_account );
         memcpy( borrowed_account->pubkey->key, &txn_ctx->accounts[0], sizeof(fd_pubkey_t) );
 
-        void * borrowed_account_data = fd_spad_alloc( txn_ctx->spad, FD_SPAD_ALIGN, FD_ACC_TOT_SZ_MAX );
+        void * borrowed_account_data = fd_spad_alloc( txn_ctx->spad, FD_ACCOUNT_META_ALIGN, FD_ACC_TOT_SZ_MAX );
```
[the remark this answers is not in the document]
Comment (yufeng-jump) [reply src/flamenco/runtime/fd_runtime.c:1826]: It's not clear to me if FD_BORROWED_ACCOUNT_ALIGN is the alignment of the fd_borrowed_account_t struct or the alignment of the data.  If it's the alignment of the struct, then it seems to have nothing to do with this allocation?  In which case we should define an alignment for the data, and take the max of that alignment and FD_ACCOUNT_META_ALIGN, and whatever alignment we care about in the allocated region.  Or some FD_STATIC_ASSERT about all the alignments involved so we are not constantly doing max(...) at run time.

[the remark this answers is not in the document]
Comment (topointon-jump) [reply src/flamenco/runtime/fd_runtime.c:1826]: Oh sorry, I misread the allocation. You're correct - `FD_BORROWED_ACCOUNT_ALIGN` is the alignment of the `fd_borrowed_account_t` struct. The allocation here is used to store the [metadata + raw data] of the borrowed account. So aligning it to the metadata is correct 😄

[the remark this answers is not in the document]
Comment (topointon-jump) [reply src/flamenco/runtime/fd_runtime.c:1826]: > In which case we should define an alignment for the data, and take the max of that alignment and FD_ACCOUNT_META_ALIGN, and whatever alignment we care about in the allocated region.

Completely agree with this

[the remark this answers is not in the document]
Comment (topointon-jump) [reply src/flamenco/runtime/fd_runtime.c:1826]: https://github.com/firedancer-io/firedancer/blob/main/src/flamenco/runtime/fd_borrowed_account.c#L63

As you say, we should also ensure this is aligned correctly.

File: src/app/fdctl/run/tiles/fd_replay.c [hunk 3]
```diff
@@ -1462,8 +1462,8 @@ unprivileged_init( fd_topo_t *      topo,
     ctx->bmtree[i]           = FD_SCRATCH_ALLOC_APPEND( l, FD_BMTREE_COMMIT_ALIGN, FD_BMTREE_COMMIT_FOOTPRINT(0) );
   }
   void * tpool_worker_mem    = FD_SCRATCH_ALLOC_APPEND( l, FD_SCRATCH_ALIGN_DEFAULT, tile->replay.tpool_thread_count * TPOOL_WORKER_MEM_SZ );
-  ulong  total_spad_size     = fd_ulong_align_up( 128UL * FD_ACC_TOT_SZ_MAX, FD_SPAD_ALIGN );
-  void * spad_mem            = FD_SCRATCH_ALLOC_APPEND( l, fd_spad_align(), tile->replay.tpool_thread_count * total_spad_size );
+  ulong  thread_spad_size    = fd_spad_footprint( 128UL * fd_ulong_align_up( FD_ACC_TOT_SZ_MAX, FD_ACCOUNT_META_ALIGN ) );
```
Comment (ibhatt-jumptrading) [inline src/app/fdctl/run/tiles/fd_replay.c:1465]: I'm pretty sure we want each borrowed allocation to be aligned with the start of an spad frame. So, we should make the footprint such that each account's allocation is aligned up to FD_SPAD_ALIGN

File: src/flamenco/runtime/fd_runtime.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (ibhatt-jumptrading) [reply src/flamenco/runtime/fd_runtime.c:1826]: The account data is laid out such that the actual account data follows the account meta. This is the same pattern used in funk

[the remark this answers is not in the document]
Comment (ibhatt-jumptrading) [reply src/flamenco/runtime/fd_runtime.c:1826]: We can just create an ifdef for FD_RECORD_DATA_ALIGN (FD_ACCOUNT_META_ALIGN) so it is more obvious

**Review by ibhatt-jumptrading — APPROVED** [review]

**Pull request merged**